core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java

```diff
@@ -306,6 +307,11 @@ public HttpRemoteTask(
             taskStatusFetcher.addStateChangeListener(newStatus -> {
                 TaskState state = newStatus.getState();
                 if (state.isDone()) {
+                    if (state == FINISHED) {
+                        // Allow dynamicFiltersFetcher to fetch DF after successful completion of task
+                        // in case of fault-tolerant execution
+                        dynamicFiltersFetcher.fetchFinalDynamicFiltersIfNecessary();
```

Comment:
Why does it have to be triggered explicitly? Shouldn't it be triggered via a normal path using the versioning mechanism (if there's a new version  - fetch it?)

Reply:
I had to do this because the existing mechanism shuts itself down after task finishes (if join itself is finished then no need to wait for DF on the probe side of the join). In the fault tolerant execution, dynamic filter collection should take place even if the build side stage is done as the probe side scan may not have finished yet.
Alternatively, I can let the existing DynamicFiltersFetcher code path keep fetching until it has the version of DF known at task finish. But that can potentially result in some extra fetches when the join completes very quickly in pipelined mode, not sure if that is worth worrying about though.

Reply:
> Basically we don't want to issue an extra fetch if a join finishes before dynamic filters are available?

Yes
It does seems to be a very corner case and not worth the added complexity.
I've simplified this now to avoid the explicit triggering. I still had to make a small change to avoid shutting down DynamicFilterFetcher from HttpRemoteTask after the task is finished.